_posts/2018-05-08-is-graphql-the-future.markdown

```diff
@@ -0,0 +1,322 @@
+under `Artwork`. There can be multiple paths to reach the same GraphQL type. This is defined explicitly by the
+schema.
+
+Usefully, the server’s response to a GraphQL request will directly mirror the shape of the request itself. The
+result of the query above looks like:
+
+```
+{
+  "data": {
+    "artwork": {
+      "title": "Campbell's Soup I: Black Bean",
+      "artist": {
+        "name": "Andy Warhol"
+      }
+    }
+  }
+}
+```
+
+# GraphQL as a (meta-)scripting language
+
+Let’s dig a little deeper into the scripting language interpretation of GraphQL, because this is the crux of how I
+think people should think of GraphQL. If I were to guess, I think Facebook…
+
+* …knows this is true. After all, much of the spec is devoted to
+  [the execution model of GraphQL](http://facebook.github.io/graphql/October2016/#sec-Execution).
+* …might have backed into this design. It’s well known that they think of their data as a graph, so I suspect
+  GraphQL might have begun literally as a "graph query language", analogous to
+  [SQL](https://en.wikipedia.org/wiki/SQL) for relational databases.
+* …thinks that this too difficult to explain, and thus, settled on the query language paradigm.
+
+There are a couple reasons GraphQL might not look like a scripting language to you. It didn’t to me, at first! After
+all, you don't write your query as sequence of operations. It doesn’t have a concept of variables, other than
+parameters to the whole document. There are no looping constructs or recursion. But I think a closer look might
+shift your perspective.
+
+## Operation flow
+
+It’s true that a GraphQL request doesn’t follow the same vertical sequence of steps model familiar to most
+programming languages. But sequencing _does_ exist. It’s just represented by calling nested fields of object types,
+terminating in a scalar field. See this query:
+
+```
+{
+  step1(arg: “something”) {
+    step2 {
+      step3(arg: "something else”) {
+        outputScalar
+      }
+    }
+  }
+```
+
+In a more traditional language, this would look more like:
+
+```
+step1(“something”)
+step2()
+return step3(“something else”)
+```
+
+So, sequencing got a bit more verbose, but it _is_ there.
+
+Interestingly, GraphQL reserves vertical stacking for something that’s an afterthought in most languages:
+_concurrency_. (Granted, there’s no way to
+[synchronize](<https://en.wikipedia.org/wiki/Synchronization_(computer_science)>) concurrent paths of execution.)
+I’m not going to quote [the spec](https://facebook.github.io/graphql/October2016/), but search it yourself, and you
+can find the word “parallel” in there several times. This design is intentional.
+
+## Variables
+
+One of the core aspects of programming is the ability to pass intermediate data around. The most basic way languages
+accomplish this with named variables. Many languages allow variables to be reassigned; some don't. GraphQL doesn’t
+have them at all! But that doesn’t mean data can’t be propagated.
```

Comment:
GraphQL _does_ have variables http://facebook.github.io/graphql/October2016/#sec-Language.Variables

Reply:
Damn. I forgot they called them that; I was thinking of them as parameters. I think my point still stands, but I'll have to distinguish binding of names as a part of the execution from GraphQL's variables, which are more like constants from an execution perspective.

Reply:
Trying to navigate around terminology with all sorts of meanings in different contexts has been one of the trickiest parts of writing this 😄